### PR TITLE
Implemented logic to add extra arguments to apply

### PR DESCRIPTION
**Pull request by gabrielelanaro**
Note this change could break existing code because meta becomes just an
extra keyword with no positional meaning.

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

File: dask/dataframe/tests/test_groupby.py [hunk 1]
```diff
@@ -112,6 +112,23 @@ def func(df):
...
+
+    pytest.raises(Exception, lambda: df.groupby('does_not_exist'))
+    pytest.raises(Exception, lambda: df.groupby('a').does_not_exist)
+    assert 'b' in dir(df.groupby('a'))
+
+    def func(df, c, d=3):
+        return df.assign(b=df.b - df.b.mean() + c * d)
+
+    assert_eq(df.groupby('a').apply(func, 1),
+              ddf.groupby('a').apply(func, 1))
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: Can we also test the keyword argument `d=` in another line?

File: dask/dataframe/tests/test_groupby.py [hunk 2]
```diff
@@ -112,6 +112,23 @@ def func(df):
...
 
 
+def test_full_groupby_apply_multiarg():
+    df = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6, 7, 8, 9],
+                       'b': [4, 5, 6, 3, 2, 1, 0, 0, 0]},
+                      index=[0, 1, 3, 5, 6, 8, 9, 9, 9])
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    pytest.raises(Exception, lambda: df.groupby('does_not_exist'))
+    pytest.raises(Exception, lambda: df.groupby('a').does_not_exist)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: Should these raise particular kinds of exceptions, like `AttributeError` ?

File: dask/dataframe/groupby.py [hunk 3]
```diff
@@ -1032,7 +1032,7 @@ def aggregate(self, arg, split_every, split_out=1):
                    split_out=split_out, split_out_setup=split_out_on_index)
 
     @insert_meta_param_description(pad=12)
-    def apply(self, func, meta=no_default):
+    def apply(self, func, *args, **kwargs):
```
Comment (martindurant) [inline dask/dataframe/groupby.py]: Is there any reasonable way to check that `args[0]` is in fact a meta and give deprecation warning? In py3 you can have positionally default argument, that would work too to avoid API breakage.

**Comment by gabrielelanaro** [comment]
@mrocklin  thanks for reviewing. 

There's also an extra issue that bugs me.

Let's say we pass a Future in args or kwargs, because you have a large argument I did have some issues especially with meta estimation. I will add a test-case to illustrate the issue and maybe we can find a way.

File: dask/dataframe/groupby.py (hunk 3 above)
Comment (gabrielelanaro) [reply dask/dataframe/groupby.py]: Precisely, this really bugs me.

But also with positially default arguments, I can totally see people do it in this way:

```
df.groupby().apply(f, arg1, arg2)
```

And fail miserably.

Another way is to make meta a required argument.

**Comment by mrocklin** [comment]
> Let's say we pass a Future in args or kwargs, because you have a large argument I did have some issues especially with meta estimation. I will add a test-case to illustrate the issue and maybe we can find a way.

Yes, I agree both that that is important, and that it may be challenging.  There are probably two challenges here:

1.  Getting the task graph of a delayed or dd.Scalar object, and merging its task graph into the computation.  This is straightforward, and has been done many other places in the codebase, but might take a bit of time if it's your first time (although you seem to be handling everything here very well :))
2.  Getting metadata.  In principle you would have to run the computation, which probably isn't feasible as a default.  I think that in this case we probably need to force the user to provide `meta=`

**Comment by gabrielelanaro** [comment]
@mrocklin I will make the test pass, for now I'll skip the metadata issue.

File: dask/dataframe/tests/test_groupby.py [hunk 4]
```diff
@@ -118,15 +118,25 @@ def test_full_groupby_apply_multiarg():
...
-
     def func(df, c, d=3):
         return df.assign(b=df.b - df.b.mean() + c * d)
 
-    assert_eq(df.groupby('a').apply(func, 1),
-              ddf.groupby('a').apply(func, 1))
+    c = 1
+    d = 2
+    c_scalar = dd.core.Scalar({'my-scalar': c}, 'my-scalar', int)
+    d_scalar = dd.core.Scalar({'my-scalar': d}, 'my-scalar', int)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: Rather than use internal API I recommend that we us something like `df.a.sum()`.  This will make this test less brittle to internal changes and make it easier for novice maintainers to understand.

**Comment by mrocklin** [comment]
Nice test!

File: dask/dataframe/tests/test_groupby.py [hunk 5]
```diff
@@ -112,6 +112,46 @@ def func(df):
...
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    def func(df, c, d=3):
+        return df.assign(b=df.b - df.b.mean() + c * d)
+
+    c = 1
+    d = 2
+
+    c_scalar = _make_scalar(c)
+    d_scalar = _make_scalar(d)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: I suggest the following instead:

```python
c_scalar = ddf.a.sum()
d_scalar = ddf.b.mean()
```

**Comment by gabrielelanaro** [comment]
I'm having some issues dealing with delayed kwargs. I believe these don't get replaced in apply_and_enforced, is there a way to accomplish this?

**Comment by mrocklin** [comment]
Have you taken a look at `dask.delayed.to_task_dask`?

```python
def to_task_dask(expr):
    """Normalize a python object and merge all sub-graphs.

    - Replace ``Delayed`` with their keys
    - Convert literals to things the schedulers can handle
    - Extract dask graphs from all enclosed values

    Parameters
    ----------
    expr : object
        The object to be normalized. This function knows how to handle
        ``Delayed``s, as well as most builtin python types.

    Returns
    -------
    task : normalized task to be run
    dask : a merged dask graph that forms the dag for this task

    Examples
    --------
    >>> a = delayed(1, 'a')
    >>> b = delayed(2, 'b')
    >>> task, dask = to_task_dask([a, b, 3])
    >>> task  # doctest: +SKIP
    ['a', 'b', 3]
    >>> dict(dask)  # doctest: +SKIP
    {'a': 1, 'b': 2}

    >>> task, dasks = to_task_dask({a: 1, b: 2})
    >>> task  # doctest: +SKIP
    (dict, [['a', 1], ['b', 2]])
    >>> dict(dask)  # doctest: +SKIP
    {'a': 1, 'b': 2}
```

**Comment by gabrielelanaro** [comment]
@mrocklin It's possible to replace the kwargs handling with a task using to_dask_task. There's however another test that conflicts with both strategy:

```
    def test_map_partitions_keeps_kwargs_in_dict():
        df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [5, 6, 7, 8]})
        a = dd.from_pandas(df, npartitions=2)
    
        def f(s, x=1):
            return s + x
    
        b = a.x.map_partitions(f, x=5)
    
>       assert "'x': 5" in str(b.dask)
```

**Comment by mrocklin** [comment]
It seems perfectly reasonable to change that test.  My guess is that the intent behind that test is that we maintain keyword arguments in the task graph directly in some easy-to-interpret way, rather than including them as closures or something else more exotic.

**Comment by mrocklin** [comment]
This looks good to me, but @jcrist might have a better eye for issues here.  Lets give him 24 hours to see if he has time to respond (he may be out of contact for the next few days).

**Comment by gabrielelanaro** [comment]
@mrocklin @jcrist  any update on this PR?

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -475,7 +478,7 @@ def map_partitions(self, func, *args, **kwargs):
         ----------
         func : function
             Function applied to each partition.
-        args, kwargs :
+        args, kwargs : Scalar, Delayed or object
```
Comment (jcrist) [inline dask/dataframe/core.py]: `args` and `kwargs` can contain these types, but aren't these types themselves. I'd remove the type note here, and move it to the description below. Something like `Arguments and keywords may contain ``Scalar``, ``Delayed``, regular python objects`.

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -3365,22 +3370,35 @@ def map_partitions(func, *args, **kwargs):
...
-    return new_dd_object(merge(dsk, *dasks), name, meta, args[0].divisions)
+    return args, dsk
 
 
 def apply_and_enforce(func, args, kwargs, meta):
     """Apply a function, and enforce the output to match meta
 
     Ensures the output has the same columns, even if empty."""
-    df = func(*args, **kwargs)
+    df = func(*args, **dict(kwargs))
```
Comment (jcrist) [inline dask/dataframe/core.py]: Why this change? I don't think this should be necessary.

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -3349,6 +3353,7 @@ def map_partitions(func, *args, **kwargs):
     name = '{0}-{1}'.format(name, token)
 
     from .multi import _maybe_align_partitions
+    args, args_dasks = _process_lazy_args(args)
```
Comment (jcrist) [inline dask/dataframe/core.py]: I *think* this can be moved down to where the call to `to_task_dask` is with no issues (since alignment also ignores `Delayed` objects). I'd prefer to group those together if possible.

File: dask/dataframe/groupby.py (hunk 3 above)
Comment (jcrist) [reply dask/dataframe/groupby.py]: I'm fine with having things fail here, people really should be using keyword arguments by name instead of by position.

File: dask/dataframe/core.py (hunk 7 above)
Comment (gabrielelanaro) [reply dask/dataframe/core.py]: It was something I forgot to clean up, I'll remove it

**Comment by jcrist** [comment]
LGTM. Thanks @gabrielelanaro! Merging.
